verification/verification_test.go

```diff
@@ -198,3 +198,59 @@ func testInitContext(d TestDPEInstance, t *testing.T) {
 		}
 	}
 }
+
+func TestCertifyKey(t *testing.T) {
+	simulators := []TestDPEInstance{
+		// No extra options besides AutoInit.
+		&DpeSimulator{exe_path: *sim_exe, supports: Support{AutoInit: true}},
+		// Supports AutoInit and simulation contexts.
+		&DpeSimulator{exe_path: *sim_exe, supports: Support{AutoInit: true, Simulation: true}},
+	}
+
+	for _, s := range simulators {
+		s.SetLocality(DPE_SIMULATOR_AUTO_INIT_LOCALITY)
+		testCertifyKey(s, t)
+	}
+}
+
+func testCertifyKey(d TestDPEInstance, t *testing.T) {
+	if d.HasPowerControl() {
+		err := d.PowerOn()
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer d.PowerOff()
+	}
+	client, err := NewClient256(d)
+	if err != nil {
+		t.Fatalf("Could not initialize client: %v", err)
+	}
+
+	certifyKeyReq := CertifyKeyReq[SHA256Digest]{
+		ContextHandle: [16]byte{0},
+		Flags:         0,
+		Label:         [32]byte{0},
+	}
+	certifyKeyResp, err := client.CertifyKey(&certifyKeyReq)
+	if err != nil {
+		t.Fatalf("Could not certify key: %v", err)
+	}
+	t.Logf("Certificate: %x", certifyKeyResp.Certificate)
+
+	if d.GetSupport().Simulation {
+		// Certify keys for each new TCI
+		for i := uint32(0); i < client.MaxTciNodes-1; i++ {
+			initCtxResp, err := client.InitializeContext(NewInitCtxIsSimulation())
+			if err != nil {
+				t.Fatal("The instance should be able to create a simulation context.")
+			}
+			defer client.DestroyContext(NewDestroyCtx(initCtxResp.Handle, false))
```

Comment:
Does this get called on each for loop iteration or when the function exits?

If it's the former, it's maybe a bit of an odd test because it's basically doing the same thing over and over again.

Also maybe a more interesting test would be to call DeriveChild then CertifyKey in each iteration. Then you'll end up with certs with different numbers of TCIs populated.

Reply:
Ah, DeriveChild is not yet supported. 😅 I've replaced this loop with a TODO.